File: packages/public-api/src/ResourceService.ts
import Fetch from "./Fetch.js";
import { handleResponse } from "./handleResponse.js";
import { Response, Problem, ResourceResponse, CreateResourceRequest, RenameResourceRequest } from "./Types.js";

export default class ResourceService {
  private readonly path = `resource`;
  constructor(private readonly fetch: Fetch) {}

  /** Creates a Resource - a non-searchable, non-filterable field (commonly a URL). */
  async create(request: CreateResourceRequest): Promise<Response<ResourceResponse, Problem>> {
    const response = await this.fetch.post(this.path, request);
    return handleResponse(response, async () => (await response.json()) as ResourceResponse);
  }

  /** Renames a Resource. Breaking change - automatically updates the List version. */
  async rename(account_name: string, list_name: string, resource_name: string, request: RenameResourceRequest): Promise<Response<ResourceResponse, Problem>> {
    const path = `${this.path}/${encodeURIComponent(account_name)}/${encodeURIComponent(list_name)}/${encodeURIComponent(resource_name)}`;
    const response = await this.fetch.put(path, request);
    return handleResponse(response, async () => (await response.json()) as ResourceResponse);
  }

  /** Removes a Resource. Breaking change - automatically updates the List version. */
  async remove(account_name: string, list_name: string, resource_name: string): Promise<Response<void, Problem>> {
    const path = `${this.path}/${encodeURIComponent(account_name)}/${encodeURIComponent(list_name)}/${encodeURIComponent(resource_name)}`;
    const response = await this.fetch.delete(path);
    return handleResponse(response, () => undefined);
  }
}

File: dotnet/SearchStack.PublicApi/Services/ResourceService.cs
using System.Threading;
using System.Threading.Tasks;
using static SearchStack.PublicApi.Url;

namespace SearchStack.PublicApi;

/// <summary>Resource create / rename / remove operations.</summary>
public sealed class ResourceService
{
    private const string Path = "resource";
    private readonly Transport _transport;

    internal ResourceService(Transport transport) => _transport = transport;

    /// <summary>Creates a Resource - a non-searchable, non-filterable field (commonly a URL).</summary>
    public Task<Response<ResourceResponse>> CreateAsync(CreateResourceRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<ResourceResponse>(Path, request, cancellationToken);

    /// <summary>Renames a Resource. Breaking change - automatically updates the List version.</summary>
    public Task<Response<ResourceResponse>> RenameAsync(string accountName, string listName, string resourceName, RenameResourceRequest request, CancellationToken cancellationToken = default) =>
        _transport.PutAsync<ResourceResponse>($"{Path}/{Enc(accountName)}/{Enc(listName)}/{Enc(resourceName)}", request, cancellationToken);

    /// <summary>Removes a Resource. Breaking change - automatically updates the List version.</summary>
    public Task<Response> RemoveAsync(string accountName, string listName, string resourceName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync($"{Path}/{Enc(accountName)}/{Enc(listName)}/{Enc(resourceName)}", cancellationToken);
}
